packages/fx-core/src/plugins/resource/aad/utils/configs.ts

```diff
@@ -213,7 +316,69 @@ export class SetApplicationInContextConfig {
       );
     }
   }
+  public restoreConfigFromLocalSettings(localSettings: v2.LocalSettings): void {
+    const frontendDomain = localSettings.frontend?.tabDomain;
+    if (frontendDomain) {
+      this.frontendDomain = format(frontendDomain as string, Formats.Domain);
+    }
+    const botId = localSettings.bot?.botId;
+    if (botId) {
+      this.botId = format(botId as string, Formats.UUID);
+    }
+
+    const clientId = localSettings.auth?.clientId;
+    if (clientId) {
+      this.clientId = clientId as string;
+    } else {
+      throw ResultFactory.SystemError(
+        GetConfigError.name,
+        GetConfigError.message(Errors.GetConfigError(ConfigKeys.clientId, Plugins.pluginName))
+      );
+    }
+  }
+  public restoreConfigFromEnvInfo(ctx: v2.Context, envInfo: v3.EnvInfoV3): void {
+    const solutionSettings = ctx.projectSetting.solutionSettings as v3.TeamsFxSolutionSettings;
+    const aadResource = envInfo.state[BuiltInResourcePluginNames.aad] as v3.AADApp;
+    let frontendDomain = aadResource?.domain;
+    if (!frontendDomain) {
+      const tabModules = solutionSettings.modules.filter((m) =>
+        m.capabilities.includes(TabOptionItem.id)
+      );
+      if (tabModules.length > 0) {
+        const hostingPlugin = tabModules[0].hostingPlugin;
+        if (hostingPlugin) {
+          frontendDomain = (envInfo.state[hostingPlugin] as v3.AzureStorage)?.domain;
+        }
+      }
+    }
+
+    if (frontendDomain) {
+      this.frontendDomain = format(frontendDomain as string, Formats.Domain);
+    }
 
+    const botModules = solutionSettings.modules.filter((m) =>
+      m.capabilities.includes(BotOptionItem.id)
+    );
+    let botId;
+    if (botModules.length > 0) {
+      const hostingPlugin = botModules[0].hostingPlugin;
```

Comment:
Can one Teams App add multiple Bots? If not, may be we don't need to use array for botModules?